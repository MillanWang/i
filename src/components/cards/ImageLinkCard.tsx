import * as React from 'react';
import {
    Accordion, AccordionDetails, AccordionSummary,
    Button,
    Card, CardActions, CardContent, CardMedia,
    Divider,
    Link,
    Typography
} from '@mui/material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';

export type LinkObject = {
    linkText: string,
    url: string
}

export function createLinkObject(linkText: string, url: string): LinkObject {
    return { linkText, url }
}

export type ImageLinkCardProps = {
    title: string,
    subtitles: string[],
    descriptionStrings: string[],
    links: LinkObject[],
    img: string,

    width?: number,
    useAccordionDescription?: boolean,
    imageHeightLimit?: number,
};

export function createImageLinkCardProps(
    title: string,
    subtitles: string[],
    descriptionStrings: string[],
    links: LinkObject[],
    img: string): ImageLinkCardProps {
    return { title, subtitles, descriptionStrings, links, img };
};


const ImageLinkCard = (
    {
        // Required but can be empty. Force developer to think abouty this
        title,
        subtitles,
        descriptionStrings,
        links,
        img,
        // Optional
        width,
        useAccordionDescription,
        imageHeightLimit,
    }: ImageLinkCardProps) => {

    return (
        <Card sx={outerCardTheme(width)} raised>

            {/* Optional top image. Only shows if non empty is given */}
            <CardImage img={img} imageHeightLimit={imageHeightLimit} />

            {/* Text Content */}
            <CardContent sx={img ? hasImageCardContentTheme : noImageCardContentTheme}>

                <Typography children={title} variant="h5" id={'Card_' + title} />

                <SubtitlesArea title={title} subtitles={subtitles} />

                <Divider sx={headerBodyDividerTheme} />

                <DescriptionStringsArea
                    cardTitle={title}
                    accordionWrap={useAccordionDescription}
                    descriptionStrings={descriptionStrings}
                />

                {/* Add bottom divider if there are non-empty links and non-empty descriptions*/}
                {links.length > 0 && descriptionStrings.length > 0 &&
                    <Divider sx={bodyLinksDividerTheme} />
                }

            </CardContent >
            {/* End Text Content */}

            {/* Links */}
            <CardActions sx={cardActionLinksTheme}>
                {links.map((linkObj: LinkObject, i: number) => (
                    <Link href={linkObj.url} target="_blank" rel="noreferrer noopener" key={'Card_' + title + '_link' + i++}>
                        <Button size="small" sx={linkButtonTheme} >
                            {linkObj.linkText}
                        </Button>
                    </Link>
                ))}
            </CardActions>
        </Card >
    )
}

type CardImageProps = {
    img: string,
    imageHeightLimit?: number,
}

const CardImage = ({ img, imageHeightLimit }: CardImageProps) => {
    if (img === '') {
        return <React.Fragment></React.Fragment>
    } else {
        return (
            <React.Fragment>
                <CardMedia
                    image={img}
                    component="img"
                    alt="Image Error"
                    height={imageHeightLimit}
                />
                <Divider sx={imageDividerTheme} />
            </React.Fragment>
        )
    }
};

type SubtitlesAreaProps = {
    title: string,
    subtitles: string[],
}

const SubtitlesArea = ({ title, subtitles }: SubtitlesAreaProps) => {
    return (
        <React.Fragment>
            {subtitles.map((subtitle: string, i: number) => (
                <Typography key={'Card_' + title + '_subtitle' + i++} variant="body2" color="text.secondary">
                    {subtitle}
                </Typography>
            ))}
        </React.Fragment>
    )
};

//TODO : There should be a separation between the desc strings alone as a bunch of typographies and then them inside an accordion. Too much for one component

type DescriptionStringsAreaProps = {
    cardTitle: string,
    accordionWrap?: boolean,
    descriptionStrings: string[]
}

const DescriptionStringsArea = ({ cardTitle, accordionWrap, descriptionStrings }: DescriptionStringsAreaProps) => {
    const descriptionStringsElement: JSX.Element = (
        <React.Fragment>
            {descriptionStrings.map((description: string, i: number) => (
                <Typography key={'Card_' + cardTitle + '_description' + i++} variant="body2" sx={descriptionStringTheme} align="left">
                    {description}
                </Typography>
            ))}
        </React.Fragment>
    );


    // If no accordion wrap, return the Typographies directly
    if (accordionWrap === undefined || !accordionWrap) { return descriptionStringsElement; }

    // Only make a state hook if accordion is used cause state is not relevant otherwise
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [accordionText, setAccordionText] = React.useState("Show Details");

    const swapString = (event: React.SyntheticEvent, expanded: boolean) => {
        expanded ? setAccordionText("Hide Details") : setAccordionText("Show Details")
    };

    return (
        <Accordion sx={accordionTheme} onChange={swapString}>
            <AccordionSummary
                sx={accordionTitleTheme}
                expandIcon={<ExpandMoreIcon htmlColor='white' />}
                children={accordionText}
            />
            <AccordionDetails
                sx={accordionDetailsTheme}
                children={descriptionStringsElement}
            />
        </Accordion>
    );
}

/******************************
 * THEMES
 *****************************/

const outerCardTheme = (width?: number,) => {
    return {
        width: width ? width : "auto",
        minWidth: 400,
        marginTop: 1,
        height: "fit-content",
        bgcolor: "primary.main",
    }
};

const hasImageCardContentTheme = {
    paddingBottom: 0,
    paddingTop: 0,
};

const noImageCardContentTheme = {
    ...hasImageCardContentTheme,
    paddingTop: 2,
}

const headerBodyDividerTheme = {
    marginBottom: 1,
    borderBottomWidth: 3,
    borderColor: "text.disabled"
};

const bodyLinksDividerTheme = {
    borderBottomWidth: 3,
    borderColor: "text.disabled"
};

const imageDividerTheme = {
    width: "95%",
    marginTop: 2,
    marginBottom: 0,
    paddingBottom: 0,
    borderBottomWidth: 3,
    borderColor: "text.disabled"
};

const descriptionStringTheme = {
    marginBottom: 1.5,
    color: "text.secondary",
}

const cardActionLinksTheme = {
    justifyContent: 'center'
};

const linkButtonTheme = {
    color: "text.primary",
    bgcolor: "primary.light",

    '&:hover': {
        backgroundColor: 'primary.dark',
    },
}

const accordionTheme = {
    backgroundColor: "primary.light",
    marginBottom: 0.75,
};

const accordionTitleTheme = {
    '&:hover': {
        backgroundColor: 'primary.dark',
    },
}

const accordionDetailsTheme = {
    backgroundColor: "primary.main",
    border: 8,
    borderColor: "primary.dark",
    marginBottom: 1,
}

export default ImageLinkCard;