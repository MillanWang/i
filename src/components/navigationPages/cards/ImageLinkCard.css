.divider {
    width: "95%";
    margin-top: 2;
    margin-bottom: 0;
    padding-bottom: 0;
    border-bottom-width: 3;
    border-color: "text.disabled"
}