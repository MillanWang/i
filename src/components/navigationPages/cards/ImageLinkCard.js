import * as React from 'react';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';


export function createImageLinkCardProps(title, timeframe, descriptionStrings, links, img) {
    return { title, timeframe, descriptionStrings, links, img };
}

export function createLink(linkText, url) {
    return { linkText, url }
}

export default function ImageLinkCard(props) {
    return (
        <Card sx={{
            maxWidth: props.maxWidth,
            margin: 1,
            height: "fit-content"
        }}
            raised
        >
            {getImage(props.imageLinkCardProps.img)}

            <CardContent sx={{ paddingBottom: 0 }}>
                <Typography variant="h5" sx={{ borderTop: 1 }}>
                    {props.imageLinkCardProps.title}
                </Typography>

                <Typography variant="body2" color="text.secondary" gutterBottom sx={{ borderBottom: 1 }}>
                    {props.imageLinkCardProps.timeframe}
                </Typography>


                {props.imageLinkCardProps.descriptionStrings.map((s) => (
                    <Typography variant="body2" color="text.secondary" gutterBottom>
                        {s}
                    </Typography>
                ))}

            </CardContent>

            {/* Links */}
            <CardActions style={{ justifyContent: 'center' }}>
                {props.imageLinkCardProps.links.map((linkObj) => (
                    <a href={linkObj.url} target="_blank" rel="noreferrer noopener">
                        <Button size="small">{linkObj.linkText}</Button>
                    </a>
                ))}
            </CardActions>
        </Card>
    )
}

function getImage(img) {
    if (img === null) {
        return <React.Fragment></React.Fragment>
    } else {
        return (
            <CardMedia
                component="img"
                alt="Image Error"
                height="140"
                image={img}
            />)
    }
}