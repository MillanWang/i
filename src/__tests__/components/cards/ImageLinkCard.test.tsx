import React from 'react';
// import { act } from 'react-dom/test-utils';
// import * as ReactDOM from "react-dom";
import { render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom'; // For .toHaveTextContent(string)
import ImageLinkCard from '../../../components/cards/ImageLinkCard'

const SAMPLE_TITLE = 'This is the Title';

describe("<ImageLinkCard />", () => {

    it('should render to snapshot with only props', () => {
        const { container } = render(<ImageLinkCard
            title=''
            subtitles={[]}
            descriptionStrings={[]}
            links={[]}
            img=''
        />
        );
        expect(container).toMatchSnapshot();
        expect(true).toBe(true);
    });

    it('should show the title', () => {
        const { container } = render(<ImageLinkCard
            title={SAMPLE_TITLE}
            subtitles={[]}
            descriptionStrings={[]}
            links={[]}
            img=''
        />
        );
        expect(container).toMatchSnapshot();

        //No duplicate titles
        const elements: HTMLElement[] = screen.queryAllByText(SAMPLE_TITLE);
        expect(elements.length).toBe(1);
        expect(elements[0]).toHaveTextContent(SAMPLE_TITLE);
    });
});