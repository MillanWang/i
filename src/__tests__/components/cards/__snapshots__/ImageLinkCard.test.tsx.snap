// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageLinkCard /> should render to snapshot with only props 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-c9zslg-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-1fnncg9-MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        id=""
      />
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1n0ge38-MuiDivider-root"
      />
      <div
        style="text-align: left;"
      />
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      style="justify-content: center;"
    />
  </div>
</div>
`;

exports[`<ImageLinkCard /> should show the title 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-c9zslg-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-1fnncg9-MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        id="This is the Title"
      >
        This is the Title
      </h5>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1n0ge38-MuiDivider-root"
      />
      <div
        style="text-align: left;"
      />
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      style="justify-content: center;"
    />
  </div>
</div>
`;
